### ISSUE TITLE
In Blazor Wasm add an option to disable/enable rendering

### ISSUE THREAD

**Issue by Andrzej-W** [dotnet/aspnetcore#14836]
### Is your feature request related to a problem? Please describe.
Here is an example of Blazor Wasm with prerendering: https://github.com/danroth27/BlazorWebAssemblyWithPrerendering. Let's run it and navigate to `/fetchdata` page. Now refresh the page in the browser. You will see 3 "different pages". First, page prerendered on the server is displayed. Then it is replaced with `Loading...`, and finally after fetch we see a complete page rendered by the application in the browser. It is annoying, and it is even more annoying in more complex application where everything is dynamic, for example application menu, and you have a few data driven components on the single page.

### Describe the solution you'd like
In application which is prerendered I would like to have an option to disable/enable rendering. Somewhere in `program.cs` or `startup.cs` I would like to call `DisableRendering()` and in the case of default application template I would like to call `EnableRendering()` after `forecast = await ...` This call should apply whole render tree to actual DOM effectively replacing prerendered content in one step. In demo application forecast is random but in typical application data used to prerender the page and then data used to render the same page in the browser will be the same. DOM update will be unnoticeable.

### Some suggestions
1. In applications like this created by default Blazor template there are some pages which are data driven (Fetch data) and some "simple" pages like Home and Counter. It is to much work to override `OnInitialized` in every page only to call `EnableRendering()` Maybe you should call `EnableRendering()` automatically in the framework just after OnInitiazed/OnInitializedAsync returns or below the place where you check if it is implemented.
2. When I'm writing "OnInitializedAsync returns" I mean it really returns, not after first await.
3. `EnableRendering()` should be called only for real pages, not for components which have `@page` directive but are used as normal components on other pages. This is necessary because we want to enable rendering only after all potentially data driven components have its data.
4. Call to OnAfterRender for each component should be delayed.

### My wish
I often hear from different people that Blazor Wasm is not for ordinary web applications and I should use MVC or Razor Pages. I have some experience with ASP.NET Core MVC and I agree that it is great technology but my opinion is different. Today almost all web applications have some interactivity and for C# people it is 100 times easier to implement this in Blazor than in JavaScript. With Blazor Wasm prerendering all main problems are solved:
- users can see first page immediately (we can even use server cacheing easily - see #14821) 
- problems with SEO are solved - search robots can see whole page content and we can dynamically render anything in `<head>` section of HTML document

This annoying page flashing when page is first displayed as prerendered, then we have nothing to display so we have "Loading..." and finally we see page rendered in the browser is probably the last problem to solve for "ordinary" web applications in Blazor Wasm. Please implement this before May 2020.

Of course there are some other problems in Blazor Wasm like support for time zones, full internationalization, etc. but they exist in every Blazor Wasm app, not only prerendered.

**Comment by javiercn** [comment]
@Andrzej-W Thanks for contacting us.

I don't think this is wasm specific, it just shows up in wasm because there is more latency. I believe this only affects the initial render and I believe if we wanted we could do a small tweak to produce a better output.

I believe you could already achieve this by overriding ShouldRender in your component.

@SteveSandersonMS One thing that would work here would be tell the Renderer on the initial render that the component was prerendered and in that case, produce a single batch in whatever of this two events happens first:
* The task for RenderComponentAsync finishes (as this waits for the initial render to be completed including async operations)
* A maximum configurable threshold time is reached (to prevent potentially waiting indefinitely).

The trade-off here is time to interactive vs flickering UI. If we delay the initial client-side render we might loose events, if we render the app immediately we have the problems @Andrzej-W described.

If we decide to delay the render, apps can always include a "Loading" banner or similar that appears on the prerendered content and disappears when the app has loaded or disable buttons and other elements during prerendering so that they only become active when the app is loaded.

Thoughts?

**Comment by pranavkm** [comment]
Thanks for the issue report to @Andrzej-W. 

We've moved this issue to the Backlog milestone. This means that it is not going to happen for the coming release. We will reassess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by Andrzej-W** [comment]
@javiercn I don't use Blazor Server Side at all and I don't know if this problem is visible there. Some time ago I read somewhere that application state in Blazor SS is preserved after prerendering and browser connects to existing circuit - it is possible that in Blazor SS this problem does not exists but as I said above I don't use it and I have no experience nor any knowledge.

Problem is visible only when you make the first call from browser to the server. If you navigate inside application it works as expected: I see "loading" and then page is rendered.

Unfortunately ShouldRender doesn't work here because component is always rendered the first time. It is documented and to be sure I checked this yesterday.

> The trade-off here is time to interactive vs flickering UI.
Prerendered page is interactive in the sense that I can click on normal links to navigate to different page. New prerendered page will be displayed in the browser. There is relatively long delay between the time prerendered page is displayed and application start. We have to load a few MB. In contrast we usually have to load only a few KB to render dynamic page in a working application. Depending on the connection speed and latency it can be between a few milliseconds and a few hundred milliseconds. In most cases it will be unnoticeable and in fact page with "loading..." text is not interactive at all :)

I have just noticed that this feature request was added to the Backlog. Taking into account that we have about 6 months to Blazor Wasm release maybe you can reconsider this.

Maybe you can at least implement two (hopefully) simple functions DisableRendering/EnableRendering.

**Comment by javiercn** [comment]
@Andrzej-W If you want to avoid the UI flickering you can persist your state inside the page and have it be re hydrated on startup inside the client in the same way that it is done in things like redux. 

That's ultimately the best solution.

**Comment by Andrzej-W** [comment]
@javiercn Thank you for your time and help. I'm not a JavaScript guy and I have no experience with redux but I think it is not what I was looking for. Fortunately I have finally found a perfect and relatively simple solution. You can read more here #14956 - there is a link to working application and updated source code. I hope you will also know the answer to one question asked in that issue.

**Comment by ladeak** [comment]
I am facing the very same issue. When I tried to trick and preload the state in the Blazor app during startup, it ran into a deadlock because of an http request awaited. Otherwise first render will render the 'Loading' which causes a huge flash.

I do believe we try to find a workaround on a workaround: the original problem I tried to solve is to become search engine friendly so bing and google can find (static) my page. Hence I enabled pre-rendering.

I wish there was a way to solve the first issue (without waiting on search engines to handle webassembly) and having a "non-static" website.

**Comment by OnurGumus** [comment]
I totally agree with there is such a problem exists. You have to give us an option to preserve the prerendered content until the actual loading is done.

**Comment by SteveSandersonMS** [comment]
The pattern I'd recommend is to put prerendered output and interactive output into separate top-level elements, so the interactive one doesn't overwrite the prerendered one.

Have the interactive one hidden by CSS until you're ready to show it, then at that point change the CSS classes so the interactive one is visible and the prerendered one is hidden. This gives you perfect control over the exact timings around when to make the switchover.

**Comment by javiercn** [comment]
Agree, this solution is simple and elegant.

**Comment by Andrzej-W** [comment]
I think it is not simple at all. Let's start from default Blazor template. We can copy FetchData.razor to Weather.razor and in Weather.razor we can remove `@page "/fetchdata"` line. Now FetchData.razor can be simplified to something like this:
```html
<div class="">
  <Weather />
</div>
<div class="">
  <Weather />
</div>
```
Now let's assume that we have some logic in FetchData.razor to assign classes to div elements and the first div is visible in prerendered output and then in the WASM application we will be able to switch classes to show only second div when it is ready. Unfortunately I don't see any improvement here because WASM application will render first div element immediately after start, and we will see `Loading ...` message. There is no way to preserve the prerendered output. The only one I have found is described here https://github.com/dotnet/aspnetcore/issues/14956

Please take into account that we can have more than one component on the page and we can also have components in the page layout. All components can read some data from database. We can also have dynamic menu system. Everything have to be correctly prerendered (SEO) and we want to eliminate flickering (end user experience). To achieve this we have to temporary disable rendering of new content. Am I missing something?

Probably the correct place to reenable rendering is in top level layout. We need an event which will be fired after all children are rendered (I mean all awaited operations are ended).

**Comment by Andrzej-W** [comment]
We can have relatively long pages in our applications and users can have slow internet connections. It is possible that user will scroll down the page before our application is fully loaded. In a perfect solution the scroll position should not change when we enable rendering.

**Comment by javiercn** [comment]
We implemented prerendering on 5.0
